tests/unit_tests/client/util/parse.cc

```diff
@@ -37,53 +37,62 @@
-    return vec;
+struct ACLFixture
+{
+  /// @brief Creates a test std::set.
+  const std::set<kovri::core::IdentHash> CreateTestSet()
+  {
+    std::set<kovri::core::IdentHash> test_set;
+    std::uint8_t to_insert1[32];
+    std::uint8_t to_insert2[32];
+    std::uint8_t to_insert3[32];
+
+    kovri::core::Base32ToByteStream("krsxg5br", 8, to_insert1, 32);  // "Test1"
+    kovri::core::Base32ToByteStream("krsxg5bs", 8, to_insert2, 32);  // "Test2"
+    kovri::core::Base32ToByteStream("krsxg5bt", 8, to_insert3, 32);  // "Test3"
+
+    test_set.insert(kovri::core::IdentHash(to_insert1));
+    test_set.insert(kovri::core::IdentHash(to_insert2));
+    test_set.insert(kovri::core::IdentHash(to_insert3));
+    return test_set;
   }
 
-  /// @brief Create CSV vector from non-CSV "test" vector
-  const std::vector<std::string> CreateCSVVector(
-      const std::vector<std::string>& csv) {
-    std::vector<std::string> vec;
-    for (auto const& field : csv)
-      vec.push_back(std::string(field + ","));
-    return vec;
+  /// @brief Create ACL list from given std::set.
+  const std::string CreateACLList(
+      const std::set<kovri::core::IdentHash>& hash_set)
+  {
+    std::string list;
+    for (auto it = hash_set.begin(); it != hash_set.end(); it++)
+      list.append(std::string((*it).ToBase32() + ","));
+    return list;
   }
 
-  /// @brief Create record from vector
-  const std::string CreateRecord(
-      const std::vector<std::string>& vec) {
-    std::string record;
-    for (auto const& field : vec)
-      record.append(field);
-    return record;
+  /// @brief Creates a string form a given std::set for comparison.
+  const std::string SetToString(
+      const std::set<kovri::core::IdentHash>& hash_set)
+  {
+    std::stringstream hash_set_stream;
+    for (auto it = hash_set.begin(); it != hash_set.end(); it++)
+      hash_set_stream << (*it);
+    return hash_set_stream.str();
   }
 };
 
 // TODO(unassigned): improve + refactor to expand test-cases
-BOOST_AUTO_TEST_CASE(ParseCSV) {
-  CSVFixture csv;
-  // Create test fixture
-  auto test_vector = csv.CreateTestVector("test", 10);
+BOOST_AUTO_TEST_CASE(ParseACLList)
+{
+  ACLFixture acl;
+  // Create test set
+  auto test_set = acl.CreateTestSet();
 
-  // Create test record to test against parsed record
-  auto test_record = csv.CreateRecord(test_vector);
+  // Create ACL list to parse
+  auto test_acl_list = acl.CreateACLList(test_set);
 
-  // Create CSV record to parse
-  auto csv_record = csv.CreateRecord(csv.CreateCSVVector(test_vector));
+  // Parse ACL list
+  auto final_set = kovri::client::ParseACLList(test_acl_list);
 
-  // Get final parsed record, should return equivalent of test fixture
-  auto final_record = csv.CreateRecord(kovri::client::ParseCSV(csv_record));
-
-  // Test against original test record
-  BOOST_CHECK_EQUAL(final_record, test_record);
+  // Test against original test set
+  BOOST_CHECK_EQUAL(acl.SetToString(final_set), acl.SetToString(test_set));
```

Comment:
This entire ACL unit-test can (and should) be thrown out and replaced with the following:

```diff
diff --git a/src/client/tunnel.cc b/src/client/tunnel.cc
index ac423ca7..28e9aa73 100644
--- a/src/client/tunnel.cc
+++ b/src/client/tunnel.cc
@@ -506,7 +506,7 @@ void I2PServerTunnel::SetACL() {
     return;
   }
   // Set ACL
-  m_ACL = kovri::client::ParseACLList(list);
+  m_ACL = kovri::client::ParseACL(list);
 }
 
 void I2PServerTunnel::Accept() {
diff --git a/src/client/util/parse.cc b/src/client/util/parse.cc
index 583eb3eb..7e7bd2e8 100644
--- a/src/client/util/parse.cc
+++ b/src/client/util/parse.cc
@@ -40,7 +40,7 @@
 
 namespace kovri {
 namespace client {
-const std::set<kovri::core::IdentHash> ParseACLList(const std::string list)
+const std::set<kovri::core::IdentHash> ParseACL(const std::string list)
 {
   std::vector<std::string> parsed;
   boost::split(parsed, list, boost::is_any_of(","));
diff --git a/src/client/util/parse.h b/src/client/util/parse.h
index f25cd8c8..983ebe61 100644
--- a/src/client/util/parse.h
+++ b/src/client/util/parse.h
@@ -43,7 +43,7 @@ namespace client {
 /// @brief Parses an ACL list
 /// @param record String record of CSV
 /// @return Returns a std::set of the b32 of each value from the list
-const std::set<kovri::core::IdentHash> ParseACLList(const std::string list);
+const std::set<kovri::core::IdentHash> ParseACL(const std::string list);
 
 /// @brief Parse for multiple CSV destination(s) and also dest:port
 /// @details Free function used solely for configuration which shows that we
diff --git a/tests/unit_tests/client/util/parse.cc b/tests/unit_tests/client/util/parse.cc                                                                                                                        
index 74e5428e..a934794e 100644                                                                                                                                                                                   
--- a/tests/unit_tests/client/util/parse.cc                                                                                                                                                                       
+++ b/tests/unit_tests/client/util/parse.cc                                                                                                                                                                       
@@ -34,65 +34,50 @@                                                                                                                                                                                               
                                                                                                                                                                                                                  
 #include "client/util/parse.h"                                                                                                                                                                                   
                                                                                                                                                                                                                  
-BOOST_AUTO_TEST_SUITE(ClientParsing)                                                                                                                                                                             
-                                                                                                                                                                                                                 
-// TODO(unassigned): improve + refactor to expand test-cases                                                                                                                                                     
-struct ACLFixture                                                                                                                                                                                                
-{                                                                                                                                                                                                                
-  /// @brief Creates a test std::set.                                                                                                                                                                            
-  const std::set<kovri::core::IdentHash> CreateTestSet()                                                                                                                                                         
-  {
-    std::set<kovri::core::IdentHash> test_set;
-    std::uint8_t to_insert1[32];
-    std::uint8_t to_insert2[32];
-    std::uint8_t to_insert3[32];
-
-    kovri::core::Base32ToByteStream("krsxg5br", 8, to_insert1, 32);  // "Test1"
-    kovri::core::Base32ToByteStream("krsxg5bs", 8, to_insert2, 32);  // "Test2"
-    kovri::core::Base32ToByteStream("krsxg5bt", 8, to_insert3, 32);  // "Test3"
-
-    test_set.insert(kovri::core::IdentHash(to_insert1));
-    test_set.insert(kovri::core::IdentHash(to_insert2));
-    test_set.insert(kovri::core::IdentHash(to_insert3));
-    return test_set;
-  }
-
-  /// @brief Create ACL list from given std::set.
-  const std::string CreateACLList(
-      const std::set<kovri::core::IdentHash>& hash_set)
-  {
-    std::string list;
-    for (auto it = hash_set.begin(); it != hash_set.end(); it++)
-      list.append(std::string((*it).ToBase32() + ","));
-    return list;
-  }
-
-  /// @brief Creates a string form a given std::set for comparison.
-  const std::string SetToString(
-      const std::set<kovri::core::IdentHash>& hash_set)
-  {
-    std::stringstream hash_set_stream;
-    for (auto it = hash_set.begin(); it != hash_set.end(); it++)
-      hash_set_stream << (*it);
-    return hash_set_stream.str();
-  }
-};
+#include <set>
 
-// TODO(unassigned): improve + refactor to expand test-cases
-BOOST_AUTO_TEST_CASE(ParseACLList)
-{
-  ACLFixture acl;
-  // Create test set
-  auto test_set = acl.CreateTestSet();
-
-  // Create ACL list to parse
-  auto test_acl_list = acl.CreateACLList(test_set);
+#include "core/crypto/rand.h"
+#include "core/router/identity.h"
+#include "core/util/log.h"
 
-  // Parse ACL list
-  auto final_set = kovri::client::ParseACLList(test_acl_list);
+BOOST_AUTO_TEST_SUITE(ClientParsing)
 
-  // Test against original test set
-  BOOST_CHECK_EQUAL(acl.SetToString(final_set), acl.SetToString(test_set));
+BOOST_AUTO_TEST_CASE(ParseACL)
+{
+  std::set<kovri::core::IdentHash> idents;
+  std::uint8_t count(3);
+  std::string acl;
+
+  // Create hashes + construct malformed ACL
+  for (std::uint8_t i(0); i < count; i++)
+    {
+      // Note: not a "real" (key-generated) ident hash
+      std::array<std::uint8_t, sizeof(kovri::core::IdentHash)> rand{{}};
+      kovri::core::RandBytes(rand.data(), rand.size());
+
+      // Create hash + insert into set
+      kovri::core::IdentHash hash(rand.data());
+      idents.insert(hash);
+
+      // Log valid b32
+      const std::string b32 = hash.ToBase32();
+      LOG(debug) << "ParseACL: " << b32;
+
+      // Construct malformed ACL delimiters
+      // TODO(unassigned): extend test for malformed ACLs?
+      acl += b32;
+      if (i != (count - 1))
+        acl += ",,,,";
+    }
+
+  // Log our malformed ACL
+  LOG(debug) << "ParseACL: " << acl;
+
+  // Parse our malformed ACL
+  auto const& parsed_idents = kovri::client::ParseACL(acl);
+
+  // Check if parser fixed the ACL and if parsed ACL matches the correct ACL
+  BOOST_CHECK(parsed_idents == idents);
 }
 
 struct TunnelFixture {
@@ -137,4 +122,4 @@ BOOST_AUTO_TEST_CASE(CatchBadClientDestination) {
   // TODO(unassigned): expand test-case (see TODO in function definition)
 }
 
-BOOST_AUTO_TEST_SUITE_END()
\ No newline at end of file
+BOOST_AUTO_TEST_SUITE_END()
```